src/main/java/core/AbstractGUI.java

```diff
@@ -66,7 +71,7 @@ protected void setFrameProperties() {
      */
     protected void updateActionButtons(AbstractPlayer player, AbstractGameState gameState) {
         if (gameState.gameStatus == Utils.GameResult.GAME_ONGOING) {
-            List<AbstractAction> actions = gameState.getActions();
+            List<AbstractAction> actions = player.forwardModel.computeAvailableActions(gameState);
```

Comment:
I think we should pass the actions into this method to avoid slowing down the GUI with extra computations. The call to this method can be traced up to the main loop in the Game class where we do have the actions calculated - it'll take a couple method signatures to be changed, but I think it might be better that way? (similarly to how they're passed to the players)

Reply:
Possible - but on this one I disagree philosophically :-). This is GUI code, so speed seems irrelevant; the time between GUI tick updates is in geological time compared to any computation of available actions, so I didn't even think about optimising this.
I may have missed something though?

Reply:
I like my GUIs speedy! But that's fine 👍